src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -237,6 +237,35 @@
       (mq/remove-queue! "localhost" "com.puppetlabs.puppetdb.commands")
       (log/info "Removed legacy queue"))))
 
+(defn initialize-schema
+  "Ensure the database is migrated to the latest version, and warn if
+  it's deprecated, log and exit if it's unsupported. We do this in a
+  single connection because HSQLDB seems to get confused if the
+  database doesn't exist but we open and close a connection without
+  creating anything."
+  [db-conn-pool product-name]
+  (sql/with-connection db-conn-pool
+    (scf-store/validate-database-version #(System/exit 1))
```

Comment:
Be glad when we finally get a chance to consider our System/exits (but not an issue for this PR).